site/content/blog/2019-11-04-unit-testing-mmctl-commands.md

```diff
@@ -0,0 +1,144 @@
+
+# Testing a Use Case
+
+For the purposes of this post, we are going to be following the test for an example case, the `mmctl user search` command. This command receives a user email argument and performs a search in the remote Mattermost instance. If a user with that email exists, it will print its information; if it doesn't, it will print an error message.
+
+This is the test function that checks the behavior when the user does exist in the remote instance:
+
+```go
+func (s *MmctlUnitTestSuite) TestSearchUserCmd() {
+	s.Run("Search for an existing user", func() {
+		emailArg := "example@example.com"
+		mockUser := model.User{Username: "ExampleUser", Email: emailArg}
+
+		s.client.
+			EXPECT().
+			GetUserByEmail(emailArg, "").
+			Return(&mockUser, &model.Response{Error: nil}).
+			Times(1)
+
+		err := searchUserCmdF(s.client, &cobra.Command{}, []string{emailArg})
+		s.Require().Nil(err)
+		s.Require().Equal(&mockUser, printer.GetLines()[0])
+		s.Require().Len(printer.GetErrorLines(), 0)
+	})
+}
+```
+
+Let's split the test and go through each one of its parts.
+
+# Adding a New Test to The Suite
+
+So first things first, we need to create the test function for our command. We use one function per command, and then we separate the different test cases with `s.Run` blocks.
+
+Our tests are part of a [`testify` suite](https://godoc.org/github.com/stretchr/testify/suite), that prepares the environment for us, and generates the mocked client so we can easily use it inside our tests. To add a new test function to the suite, we just need to define it on the suite struct:
+
+```go
+func (s *MmctlUnitTestSuite) TestSearchUserCmd() {
+    s.Run("Our test case", func() {
+        // ...
+    })
+}
+```
+
+This way, we will have an `s` instance inside our test function that will contain a mocked instance of the client ready for us to use, and we will be able to use the suite to run our assertions.
+
+# Mocking an Interaction With the Server
+
+So now that we have our test function defined, the next step is to think about our test case: what inputs is the command going to receive? What interactions with the server are those inputs going to cause? And what responses do I want to mock for them?
+
+For our given test case, we are going to receive an email address as the input, and we want to mock a valid response for it. The client method that our command is going to use to send the request to the server is the `GetUserByEmail` method, and for the case of an existing user, it will return the found user instance:
```

Comment:
```suggestion
For our given test case, we are going to receive an email address as the input and we want to mock a valid response for it. The client method that our command is going to use to send the request to the server is the `GetUserByEmail` method, and for the case of an existing user, it will return the found user instance:
```